https://thefiddler.substack.com/p/happy-almost-new-year-from-the-fiddler-f0b

The number 2025 is not prime. As a matter of fact, it’s a perfect square: 2025 = 452.

You cannot make 2025 by adding two distinct primes. To do so, you’d have to add an even prime and an odd prime. The only even prime is 2, but 2025 − 2 = 2023, which is not prime (it’s equal to 7∙172).

But you can make 2025 by adding three distinct primes. For example, 661 + 673 + 691 = 2025.

You can also make 2025 by adding four distinct primes: 2 + 659 + 673 + 691 = 2025.

**What is the greatest number of distinct primes that add up to 2025?**

# TLDR: 32, see bottom cell

Let's try the programmatic approach, although given the combinatorial explosion, I'm expecting it will not scale
If not, let's see if I can use some properties of prime numbers to either solve the problem mathematically, or to reduce the search space

In [3]:
from itertools import combinations
from math import comb

def is_prime(n:int)->bool:
    if n <= 1:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

primes = [num for num in range(2025) if is_prime(num)]
print(f"{len(primes)} primes under 2025:\n{primes}")


306 primes under 2025:
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 119

What's the upper bound, ie the max number of primes that could possibly add up to 2025?

In [6]:
nb=0
s=0
for prime in primes:
    if s+prime>2025:
        break
    s+=prime
    nb+=1
print(f"the first {nb} primes add up to {s}")
print(f"the first {nb+1} primes add up to {s+prime}")

the first 33 primes add up to 1988
the first 34 primes add up to 2127


OK we have a narrower scope now, although this will likely represent an untractable search space

In [7]:
def get_relevant_primes(cardinal:int, primes:list[int])->list[int]:
    """Returns the list of candidate primes in which one might extract `cardinal`
    numbers to add up to 2025; based on the max plausible value.

    Args:
        cardinal (int): the number of primes in the group of candidate numbers

    Returns:
        list[int]: the group of candidate primes
    """
    s= sum(primes[:cardinal-1]) # take the `cardinal-1` first primes
    # what is the highest prime that could be taken, ie such that s + this prime <= 2025 ?
    for i in range(cardinal-1, len(primes)):
        if s + primes[i] > 2025:
            candidates= primes[:i]
            return candidates
    return primes


In [15]:
max_cardinal = 3 # the max known (so far) number of primes adding up to 2025 provided in the text is 4
nb_combs= {}
for group_cardinal in range(max_cardinal,33):
    #print(f"{group_cardinal=}")
    candidates = get_relevant_primes(group_cardinal, primes)
    #print(f"{len(candidates)=}")
    nb_combs[group_cardinal]= comb(len(candidates), group_cardinal)
    
print("below are the number of combinations of k primes taken in the above list of 306 primes under 2025.\nAnd that is already optimized by looking at the max plausible value (see function get_relevant_primes)! As expected these are astronomical.")
print("The only value that is actually iterable for a mundane computer is 3: looping over 4_728_720 items is generally doable.")

print(sorted(nb_combs.items(), key=lambda x :x[1]))
print("These are astronomical numbers, there must be another way")

below are the number of combinations of k primes taken in the above list of 306 primes under 2025.
And that is already optimized by looking at the max plausible value (see function get_relevant_primes)! As expected these are astronomical.
The only value that is actually iterable for a mundane computer is 3: looping over 4_728_720 items is generally doable.
[(3, 4728720), (4, 353518180), (5, 20932912560), (6, 1002319312995), (7, 39494993171634), (8, 1365332543512245), (9, 41221140106119260), (32, 450883717216034179), (10, 1100220888889816408), (11, 25225723748833593440), (12, 454114140096308188315), (13, 8251293359737707230700), (31, 59488706401523551644624), (14, 128605083186263852143800), (15, 1609484153977526457766689), (16, 17811576069083142710034000), (30, 29372339821610944823963760), (17, 174573373849880188322437200), (18, 1214402797397737883192882625), (29, 4243634690199352227365018524), (19, 7192441475404230387885398800), (20, 32991718972148235529162365510), (28, 719470617146570

Now looking at the text again; as is well-known, 2 is the only even prime, and the text suggests there might be a pattern: 2 might be necessary
to build any even-length combination, seen as 'even + odd' where odd is the sum of all the primes but 2.

With this in mind, and knowing that when creating the combinations, they'll be ordered, i.e. those containing 2 will pop up first,
this gave me the idea of trying out every even-length combination, it *should* be relatively quick

In [19]:
%%time
# iterate over even cardinals => 2 is needed and this will quickly yield tuples that add up to 2025
for cardinal in range(6, 34, 2):
    candidates = get_relevant_primes(cardinal, primes)
    for combi in combinations(candidates, cardinal):
        if sum(combi)==2025:
            print(f"{'+'.join(map(str,combi))}=2025 ({cardinal} numbers)")
            break

2+3+5+7+11+1997=2025 (6 numbers)
2+3+5+7+11+13+53+1931=2025 (8 numbers)
2+3+5+7+11+13+17+19+41+1907=2025 (10 numbers)
2+3+5+7+11+13+17+19+23+29+73+1823=2025 (12 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+1787=2025 (14 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+1697=2025 (16 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+61+1583=2025 (18 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+71+1453=2025 (20 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+79+1307=2025 (22 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+1151=2025 (24 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+109+953=2025 (26 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+761=2025 (28 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+113+541=2025 (30 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+109+1

This was instantaneous (see run time above)!
What a change from the previous approach!

The answer is therefore: a maximum of **32** prime numbers can add up to 2025. As seen previously, 33 does not work as the first 33 primes add up to 1988 while the first 34 add up to 2127.
One such combination is shown above


Here are more 10 more solutions:

In [22]:
%%time

c = 0
candidates = get_relevant_primes(32, primes)
combis = combinations(candidates, 32)
while c <10:
    combi = next(combis)
    if sum(combi)==2025:
        print(f"{'+'.join(map(str,combi))}=2025 ({32} numbers)")
        c += 1

2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+109+113+139+293=2025 (32 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+109+113+149+283=2025 (32 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+109+113+151+281=2025 (32 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+109+113+163+269=2025 (32 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+109+113+181+251=2025 (32 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+109+113+191+241=2025 (32 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+109+113+193+239=2025 (32 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83+89+97+101+103+107+109+113+199+233=2025 (32 numbers)
2+3+5+7+11+13+17+19+23+29+31+37+41+43+47+53+59+61+67+71+73+79+83

Interestingly, after finding the solution, I copy-pasted the problem's text to the AI assistant I use (Copilot), and it failed to solve it